### PR TITLE
Provider rejects an application via provider_interface

### PR DESCRIPTION
**Pull request by mnacos**
### Context

The application already provides a mechanism for rejecting applications, through the API. No such capability exists yet within the provider interface. This PR adds the relevant user interface components to allow providers to reject applications within the app.

A ```RejectApplication``` service already exists in the codebase, and it is currently used by ```vendor_api/decisions_controller.rb```. The provider interface should make use of the same service to change the application's state.

### Changes proposed in this pull request

This pull request introduces a new controller ```provider_interface/decisions_controller.rb``` with these actions:

- ```#respond```
- ```#submit_response```
- ```#new_reject```
- ```#confirm_reject```
- ```#create_reject```

Small changes have been made to ```app/views/provider_interface/application_choices/show.html.erb``` to add a status box, which contains a 'Respond to application' button/link.

image

This takes you to a 'Respond to application' interface.

image

Selecting 'Reject application' takes you to the rejection form.

image

Submitting the form takes you to a confirmation step.

image

Once the rejection is complete, the user is redirected to the application #show with a status change notice.

image

### Guidance to review

1. Visit /provider/applications and click on an application to view it.
2. Try the flow described above and reject the application.
3. Run ```bundle exec rspec spec/system/provider_interface/provider_rejects_application.rb```
4. Review source code changes

### Link to Trello card

[1253 - Provider can reject an applicant](https://trello.com/c/aLWDIdpl)

### Env vars

No new environment variables.

- [x] If this PR introduces new environment variables, they have been added to all necessary parts of the Azure config based on the [documentation in the README](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -164,6 +164,10 @@
 
     get '/applications' => 'application_choices#index'
     get '/applications/:application_choice_id' => 'application_choices#show', as: :application_choice
+    get '/applications/:application_choice_id/respond' => 'application_choices#respond', as: :application_choice_respond
+    get '/applications/:application_choice_id/reject' => 'application_choices#new_reject', as: :application_choice_new_reject
+    get '/applications/:application_choice_id/reject/confirm' => 'application_choices#new_confirm_reject', as: :application_choice_new_confirm_reject
```
Comment (tijmenb) [inline config/routes.rb]: This will be better as a post to avoid PII in URLs.

File: app/controllers/provider_interface/application_choices_controller.rb [hunk 2]
```diff
@@ -10,12 +10,22 @@ def index
     end
 
     def show
-      application_choice = GetApplicationChoicesForProvider.call(provider: current_user.provider)
-        .find(params[:application_choice_id])
-
       @application_choice = ApplicationChoicePresenter.new(application_choice)
     end
 
+    def respond
```
Comment (tijmenb) [inline app/controllers/provider_interface/application_choices_controller.rb]: As discussed offline, perhaps move into a `DecisionsController` to shadow the API controller structure.

File: app/views/provider_interface/application_choices/respond.html.erb [hunk 3]
```diff
@@ -0,0 +1,48 @@
+<%= form_tag provider_interface_application_choice_new_reject_path, method: :get do %>
```
Comment (tijmenb) [inline app/views/provider_interface/application_choices/respond.html.erb]: We can use the structure from https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/blob/2fdef039740a86bc2345be7a26498bb94c4c8a5f/app/views/candidate_interface/start_page/eligibility.html.erb for this.

File: app/views/provider_interface/application_choices/new_reject.html.erb [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    <h1 class="govuk-fieldset__heading">
+      
+    <span class="govuk-caption-xl">Application for Upper Primary with Modern Foreign Languages (XR9C)</span>
+    Reject application
+  
+    </h1>
+  
+  </legend>
+ 
+<dl class="govuk-summary-list govuk-!-margin-bottom-8">
```
Comment (tijmenb) [inline app/views/provider_interface/application_choices/new_reject.html.erb]: We can use the `SummaryCardComponent` for this probably:

https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/blob/37da4915ad1ae9762e5e7b449066b1ecb9cbf41c/app/components/summary_card_component.rb

File: app/controllers/provider_interface/application_choices_controller.rb [hunk 5]
```diff
@@ -9,21 +11,13 @@ def index
...
 
   private
 
-    # Stub out the current user and their organisation. Will be replaced
-    # by a proper ProviderUser when implementing Signin.
-    def current_user
-      fake_user_class = Struct.new(:provider)
-      fake_provider_class = Struct.new(:code)
-      fake_user_class.new(fake_provider_class.new('ABC'))
+    def set_application_choice
```
Comment (tijmenb) [inline app/controllers/provider_interface/application_choices_controller.rb]: This should set an instance of `ApplicationChoicePresenter`.

File: app/controllers/provider_interface/application_choices_controller.rb [hunk 6]
```diff
@@ -1,5 +1,7 @@
 module ProviderInterface
   class ApplicationChoicesController < ProviderInterfaceController
+    before_action :set_application_choice
```
Comment (tijmenb) [inline app/controllers/provider_interface/application_choices_controller.rb]: Should not apply to `index` (or revert this controller change).

File: app/views/provider_interface/application_choices/new_reject.html.erb (hunk 4 above)
Comment (mnacos) [reply app/views/provider_interface/application_choices/new_reject.html.erb]: Works great. Changed it to support a borderless version

File: app/controllers/provider_interface/application_choices_controller.rb (hunk 5 above)
Comment (mnacos) [reply app/controllers/provider_interface/application_choices_controller.rb]: fixed

File: app/controllers/provider_interface/application_choices_controller.rb (hunk 6 above)
Comment (mnacos) [reply app/controllers/provider_interface/application_choices_controller.rb]: fixed

File: app/components/error_summary_component.html.erb [hunk 7]
```diff
@@ -0,0 +1,21 @@
+<% # see https://design-system.service.gov.uk/components/error-summary/ %>
+<% content_for :title, flush: true do %>
+  Error: <%= content_for(:title) %>
+<% end %>
+
+<div class="govuk-error-summary" aria-labelledby="error-summary-title" role="alert" tabindex="-1" data-module="govuk-error-summary">
```
Comment (tijmenb) [inline app/components/error_summary_component.html.erb]: We shouldn't use `govuk-error-summary` for flash messages, it's meant to be used for forms. Forms already have the error summary via the formbuilder.

File: app/controllers/provider_interface/decisions_controller.rb [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    def submit_response
+      Rails.logger.warn params.inspect
+      decision = params[:application_choice][:decision] if params[:application_choice]
+      if decision == 'offer'
+        redirect_to action: :new_offer
+      elsif decision == 'reject'
+        redirect_to action: :new_reject
+      else
+        raise 'Unexpected application_choice decision'
```
Comment (RobertM905) [inline app/controllers/provider_interface/decisions_controller.rb]: Should there be some handling if a provider clicks continue on the `/provider/applications/{id}/respond` page without first selecting an option?

Currently we just get an exception.

File: app/components/error_flash_component.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+  validates :message, presence: true
+
+  def initialize(message:)
+    @message = message
+  end
+
+private
+
+  attr_reader :message
+end
```
Comment (RobertM905) [inline app/components/error_flash_component.rb]: Could the flash component code be moved to a separate PR?

Feels like this could be a separate card and would be nicer in it's own PR to get Design feedback.

Comment (mnacos) [reply app/components/error_flash_component.rb]: I've already asked for design feedback on this specific story/PR. If they say it's ok I'll keep the SuccessFlashComponent because the story requires this feedback at the last stage. I'll prepare a separate PR anyway implementing different severity levels, anyway.

File: app/controllers/provider_interface/decisions_controller.rb (hunk 8 above)
Comment (mnacos) [reply app/controllers/provider_interface/decisions_controller.rb]: No, this code is actually part of ```#submit_response```, which you only get to via a POST. You'd have to post to this url without specifying 'offer' or 'reject' to get the error.

Comment (RobertM905) [reply app/controllers/provider_interface/decisions_controller.rb]: Yes but when you first navigate to the `respond` page neither option is selected. 

Screenshot 2019-11-15 at 11 56 01

And when you click continue without selecting an option you get an exception.

Screenshot 2019-11-15 at 11 57 54

Comment (mnacos) [reply app/controllers/provider_interface/decisions_controller.rb]: Oh I see. I'll remove the exception for now and return no content. Once we have a warning flash component I'll add a warning to the page.

**Comment by mnacos** [comment]
> I'm happy to approve

Great, thanks. Sorry @tijmenb I will merge this before you have time to formally approve, as Rob needs this merged in order to work on https://trello.com/c/QP87Ig3k.
